### PR TITLE
Tanium: initial merge commit

### PR DESCRIPTION
**Pull request by ian-hands-tanium**
Initial PR for Tanium Sentinel integration

### CODE REVIEW

File: Solutions/Tanium/Workbooks/TaniumWorkbook.json
Comment (v-sabiraj) [inline Solutions/Tanium/Workbooks/TaniumWorkbook.json:54]: @ian-hands-tanium, there are several tables used for workbook queries like TaniumComplyVulnerabilities_CL, TaniumComplyCompliance_CL,TaniumSentinelApplicablePatches_CL,TaniumSentinelApplicablePatches_CL,TaniumSentinelHighUptime_CL etc. Please add sample data for those too, thanks.

**Review by v-sabiraj — CHANGES_REQUESTED** [review]

**Comment by v-sabiraj** [comment]
@ian-hands-tanium, please move the logo to[ logo ](https://github.com/Azure/Azure-Sentinel/tree/master/Logos) folder and the format of the logo should be svg.

File: Solutions/Tanium/Workbooks/TaniumWorkbook.json
Comment (ian-hands-tanium) [reply Solutions/Tanium/Workbooks/TaniumWorkbook.json:526]: Thanks!

**Comment by ian-hands-tanium** [comment]
> @ian-hands-tanium, please move the logo to[ logo ](https://github.com/Azure/Azure-Sentinel/tree/master/Logos) folder and the format of the logo should be svg.

Got it, thanks!

File: Solutions/Tanium/Workbooks/TaniumWorkbook.json
Comment (ajcollins0) [reply Solutions/Tanium/Workbooks/TaniumWorkbook.json]: @v-sabiraj sample data has been added. Sample data for TaniumSentinelApplicablePatches_CL was not added because it was removed our workflows for now.

File: Solutions/Tanium/Workbooks/TaniumWorkbook.json
Comment (ajcollins0) [reply Solutions/Tanium/Workbooks/TaniumWorkbook.json]: Added this fix in the most recent commit sir!

File: Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml
Comment (ajcollins0) [reply Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml:12]: @v-sabiraj added sample data for this comment also

**Comment by harmankaur-tanium** [comment]
Request status update.

File: Solutions/Tanium/Workbooks/TaniumWorkbook.json
Comment (v-sabiraj) [inline Solutions/Tanium/Workbooks/TaniumWorkbook.json:788]: @ian-hands-tanium, A workbook template must not contain any references to resources, please update, thanks.

**Review by v-sabiraj — CHANGES_REQUESTED** [review]

**Comment by v-sabiraj** [comment]
@ian-hands-tanium, what type of files are added in data folder?

**Comment by v-sabiraj** [comment]
@devikamehra , can you please check playbooks, thanks.

File: Solutions/Tanium/Workbooks/TaniumWorkbook.json
Comment (ajcollins0) [reply Solutions/Tanium/Workbooks/TaniumWorkbook.json]: @v-sabiraj the "FallbackResourceIDs" has been removed. Please let me know if there is anything else needed, thank you!

File: Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json
Comment (devikamehra) [inline Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json:127]: Connection is not present in the ARM Template. Deployment failed and thus couldn't proceed with further testing.
Same is true for other playbooks as well.

File: Solutions/Tanium/Playbooks/Tanium-QuarantineHosts/azuredeploy.json
Comment (devikamehra) [inline Solutions/Tanium/Playbooks/Tanium-QuarantineHosts/azuredeploy.json:131]: Readme is missing for all the playbooks. Can we please add a readme for each playbook with required details. 

If we require one single common README, please include the same in Playbooks folder.

**Review by devikamehra — CHANGES_REQUESTED** [review]

File: Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml
Comment (aprakash13) [inline Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml:3]: Do we have a Typo here: Thread vs Threat

File: Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml
Comment (aprakash13) [inline Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml:9]: Please also include a relevantTechniques field.

File: Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml
Comment (aprakash13) [inline Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml:4]: Please also add the requiredDataConnectors field.

**Review by aprakash13 — CHANGES_REQUESTED** [review]
Overall, the query looks good. However, I have added a few suggestions that needs to be included in the template. 

Also, since TaniumThreatResponse_CL is a custom table, we need to define its schema. Here is an example: https://github.com/Azure/Azure-Sentinel/blob/master/.script/tests/KqlvalidationsTests/CustomTables/CarbonBlackEvents_CL.json

We need to resolve all the validation errors before this can be merged. Hopefully, with all these suggested changes the validation errors would go away in the next run. Thanks.

File: Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json
Comment (ian-hands-tanium) [reply Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json:127]: Here is what we do:
https://github.com/tanium/Azure-Sentinel/blob/tanium-dev/tanium_scripts/fix_maintemp.sh

After the build step we modify the `mainTemplate.json` a bit because there are somethings that we cannot change via the input.json. I could not figure out how to add these by tweaking the input.json...
https://github.com/tanium/Azure-Sentinel/blob/tanium-dev/tanium_scripts/fix_maintemp.sh#L14-L16

**Comment by ajcollins0** [comment]
> @ian-hands-tanium, what type of files are added in data folder? Are they Custom connector or data connector?

@v-sabiraj These are API Connections. We noticed that when we create a playbook, it needs to have an API Connection which gives it permission to access that data within Sentinel. When we were testing importing this content pack, if we didn't have one of these, all playbooks would not work. 

Here is an example: from the web GUI: 
image

File: Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml
Comment (ajcollins0) [reply Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml]: Thank you, I fixed this in the most recent commit here: 
https://github.com/Azure/Azure-Sentinel/pull/5307/commits/f581d295ef400200dc54ec77e8a279ba05bf8170

File: Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml
Comment (aprakash13) [reply Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml:4]: @ajcollins0 Please put the field here and leave it blank. Similar to what you have done for tactics.

File: Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml
Comment (aprakash13) [reply Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml:9]: Please include the field and leave it blank similar to tactics.

File: Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml
Comment (ajcollins0) [reply Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml:4]: added - thank you!

File: Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml
Comment (ajcollins0) [reply Solutions/Tanium/Analytic Rules/TaniumThreatResponseAlerts.yaml:9]: added - thank you!

**Comment by aprakash13** [comment]
Please fix the validation errors.  One step towards that would be:
Since TaniumThreatResponse_CL is a custom table, we need to define its schema. Here is an example: https://github.com/Azure/Azure-Sentinel/blob/master/.script/tests/KqlvalidationsTests/CustomTables/CarbonBlackEvents_CL.json

File: Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json
Comment (devikamehra) [reply Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json:127]: There are few things we need to do here:
1. Each playbook should have the required connections defined within their respective ARM Templates. There is no need to put them separately in Data folder. (See [CiscoUmbrella](https://github.com/Azure/Azure-Sentinel/blob/master/Solutions/CiscoUmbrella/Package/mainTemplate.json) for an example)
2. Once #1 is completed, you can proceed with running the packing tool using the input.json file.
3. There should not be any **Data** folder as mentioned above with the connections. There is always a **Data Connector** folder with the required data connector. In this case, I can see there is a Logic App based connector, is my understanding correct? If yes, connection(s) are missing from this ARM Template as well. Please add the connection within the workflow ARM Template to avoid any issues during the deployment of the connector. I understand that this would bring in duplicates resources of same type in your mainTemplate.json file but that is totally acceptable.

File: Solutions/Tanium/Playbooks/Tanium-QuarantineHosts/azuredeploy.json
Comment (devikamehra) [reply Solutions/Tanium/Playbooks/Tanium-QuarantineHosts/azuredeploy.json:131]: @ajcollins0 Thanks for adding the README files. Just a suggestion, can we look at the possibility of combining 2 or more playbooks into one single playbook unless there is an explicit reason to keep them separate.

File: Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json
Comment (ajcollins0) [reply Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json:127]: @devikamehra I'm a little confused by points 1 and 2 in your response. It looks like our maintemplate.json has the connections defined. So maybe we should simply delete the `AzureSentinelConnector.json` , `TaniumSentinelConnector.json`, and `APIConnectors.json`. 

Is there anyway we could meet and screen share to discuss further? 


For point 3, we are sending data into Log Analytics from our platform, so we don't have a Data Connector like how other venders do. 

For point 4, That is a logic app, but it's not defined as a playbook within the azure portal because its not triggered from an incident. For example, all of the other playbooks are triggered from an incident as shown here: 
image

The file in question (https://github.com/Azure/Azure-Sentinel/blob/5307dff2282d94fe716f1b2e08aafaef38597084/Solutions/Tanium/Data/AddCommentLA.json) is a logic app that has a different trigger as shown here: 
image

**Comment by ajcollins0** [comment]
> Please fix the validation errors. One step towards that would be: Since TaniumThreatResponse_CL is a custom table, we need to define its schema. Here is an example: https://github.com/Azure/Azure-Sentinel/blob/master/.script/tests/KqlvalidationsTests/CustomTables/CarbonBlackEvents_CL.json

Just added the schemas - going to see if this builds and fix if needed, thank you! 

Note: This is a holiday weekend in the US, so I may not be back working on this until Tuesday July 5th.

**Comment by ajcollins0** [comment]
@v-sabiraj @devikamehra @aprakash13 all validation errors are fixed now. Let me know what you would like to see next and I hope all is well!

**Comment by aprakash13** [comment]
> @v-sabiraj @devikamehra @aprakash13 all validation errors are fixed now. Let me know what you would like to see next and I hope all is well!

Thanks @ajcollins0. Looks good to me from a query perspective.  @v-sabiraj @devikamehra Please feel free to merge if other components look good.

**Comment by ajcollins0** [comment]
this morning I had a colleague try to import this content pack into his personal Azure account and he saw the below error, is this expected? (Note: we attempted to import via the .zip file inside the path: `Solutions/Tanium/Package/1.0.1.zip`. 
```
{
  “code”: “InvalidTemplate”,
  “message”: “Deployment template validation failed: ’The template variable ‘tanium-sentinel-connection’ is not found. Please see https://aka.ms/arm-template/#variables for usage details.’.”
}
```

**Comment by harmankaur-tanium** [comment]
Requesting an update per customer request.

File: Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json
Comment (devikamehra) [reply Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json:127]: @ajcollins0 
With point 1, what I meant is that the deployment of the playbook should be successful. When I deploy any one of the playbooks, example [Tanium-AddComment](https://raw.githubusercontent.com/Azure/Azure-Sentinel/07ac54eda970d0a464c2f0816edc749d08b9a877/Solutions/Tanium/Playbooks/Tanium-AddComment/azuredeploy.json), I get the following error.

image

This is because, the ARM Template has the dependsOn clause for the resource which does not exist in the ARM template. We have to choose from the following two options:

1. We ask the user to individually deploy this connection before-hand and then install playbook without this dependsOn clause.
2. **This is the recommended option.** Include the connection resource for the deployment within the ARM template itself. Whenever the user deploys this ARM Template, they will get both the logic app and the connection.

Please let me know in case more information is needed here.

In point 2, there should be re-packaging of the solution with the updates made in point 1.

In point 3 and 4, can you please explain how are we getting the data into the desired tables? Also, for the file in question, what is the operation we are applying using this logic app? This would help us define whether we should consider this as a playbook or a data connector or none.

File: Solutions/Tanium/Workbooks/TaniumWorkbook.json
Comment (v-sabiraj) [reply Solutions/Tanium/Workbooks/TaniumWorkbook.json]: @ajcollins0, its fine to add with empty array

**Comment by v-sabiraj** [comment]
@ajcollins0, can you please use a single logo consistent and also please check the format of logo should be .svg. Please add the logo in[ logo folder](https://github.com/Azure/Azure-Sentinel/tree/master/Logos), you can refer the other logos from the same folder, thanks.

**Comment by ajcollins0** [comment]
> > Hey @ian-hands-tanium, @ajcollins0, thanks for adding sample data. As the sample data column names are appended with _s, the data is been ingested by _s_s, can you please update sample data by just column name, e.g "Count" rather than "Count_s", thanks. image image
> 
> Can you please implement this change to all the columns added in the tables, thanks.

okay that sounds good! fixing now in the csv files. Quick Question, do you want me to remove the `_t` from any timestamps? For example, this file: https://github.com/tanium/Azure-Sentinel/blob/master/Sample%20Data/Custom/TaniumThreatResponse_CL.csv has a few columns like `Timestamp_t [UTC]`, `Match_Details_finding_first_seen_t [UTC]` `Match_Details_finding_last_seen_t [UTC]`. 

For these columns should I remove the `_t`, or the `_t [UTC]`, or should I do nothing? 

Additionally, should I edit all of the json files that are in this folder too? (note there are a few that exist in this folder, I'm only posting 1 as an example): 
https://github.com/tanium/Azure-Sentinel/blob/master/.script/tests/KqlvalidationsTests/CustomTables/TaniumComplyCompliance_CL.json

File: Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json
Comment (ajcollins0) [reply Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json:127]: @devikamehra  can you send me an example of how I can include the connection resource inside the Arm template? Every playbook should rely on the same API Connection object, so I will need to ensure each playbook has the same reference to this connection. Once I get this information I will include it and repackage the solution. 


`can you please explain how are we getting the data into the desired tables?` 
Our software has a way to ship data from our product into other products. It's called Tanium Connect. The general documentation is found here: https://docs.tanium.com/connect/connect/index.html. We are releasing a Microsoft Log Analytics connector where the customers can input their Azure information and we will send data from our platform directly into log analytics. The underlying API that we are using to send out data into Log Analytics is documented here: https://docs.microsoft.com/en-us/azure/azure-monitor/logs/data-collector-api.

`Also, for the file in question, what is the operation we are applying using this logic app? This would help us define whether we should consider this as a playbook or a data connector or none.`
I am assuming you're referring to this file? https://raw.githubusercontent.com/Azure/Azure-Sentinel/07ac54eda970d0a464c2f0816edc749d08b9a877/Solutions/Tanium/Playbooks/Tanium-AddComment/azuredeploy.json
If so, this playbook is responsible for adding specific comments into an incident. Here is the photo of the logic-app designer, this should layout what it does. It simply takes in information, then posts the information into an incident inside of sentinel. 
image

Comment (ajcollins0) [reply Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json:140]: @devikamehra I added the connection as a resource in all the playbooks, and re-ran the packaging tool to create a 1.0.2.zip file. When I try to import the maintemplate.json into the webportal  I am seeing this error:
 
"Deployment template validation failed: 'The resource 'Microsoft.Web/connections/taniumsentinel' at line '262' and column '9' is defined multiple times in a template. Please see https://aka.ms/arm-template/#resources for usage details.'. (Code: InvalidTemplate)" 

Is there a way I can use a single connection and share it for all the playbooks?

**Comment by ajcollins0** [comment]
@devikamehra my most recent push today fixed the validation issues with the mainTemplate - I think we're good to go regarding the problems you previously mentioned with the API Connections

File: Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json
Comment (devikamehra) [reply Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json]: @ajcollins0 I tried deploying the mainTemplate.json file and it is working fine for me. But you should be using v2 of the packaging tool as we are moving all solutions from v1 to v2. Let me know if you need help in packaging.

As I see that the data lands into sentinel using push base mechanism, we would need a data connector to provide uniform experience to the sentinel user. This data connector will contain steps to configure data logging from Tanium Connector. Please use this as an example to develop the data connector - https://github.com/Azure/Azure-Sentinel/blob/master/Solutions/Sophos%20Cloud%20Optix/Data%20Connectors/Connector_REST_API_SophosCloudOptix.json

Kindly follow the guidance mentioned here to develop the data connector - https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/ReadMe.md

Place the developed JSON file in **Data Connector** folder (within Solution folder) and update the input file for packaging. Use the v2 of the packaging tool to develop the package. - https://github.com/Azure/Azure-Sentinel/tree/master/Tools/Create-Azure-Sentinel-Solution/V2

Comment (ajcollins0) [reply Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json]: Everything is working and getting the green light with the most recent push, and its using V2 of the packaging tool now. But when I attempt to import the mainTemplate.json, I am given this error and I don't know what this means. 


image

Comment (ajcollins0) [reply Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json]: I should also note: Test-AzTemplate works on mainTemplate.json too, and the github actions have all passed.

File: Solutions/Tanium/Data Connectors/Connector_REST_API_Tanium.json
Comment (devikamehra) [inline Solutions/Tanium/Data Connectors/Connector_REST_API_Tanium.json:20]: Include all the data types that we are receiving from the source.

File: Solutions/Tanium/Data Connectors/Connector_REST_API_Tanium.json
Comment (devikamehra) [inline Solutions/Tanium/Data Connectors/Connector_REST_API_Tanium.json:68]: Typo:
Contact your Tanium representative to **know** about

File: Solutions/Tanium/Data Connectors/Connector_REST_API_Tanium.json
Comment (devikamehra) [inline Solutions/Tanium/Data Connectors/Connector_REST_API_Tanium.json:96]: Can we change this to bullets?

**Review by devikamehra — CHANGES_REQUESTED** [review]

**Comment by devikamehra** [comment]
Sorry for the confusion @ajcollins0. Please proceed with v1 only.

For the failing deployment, can you try deploying in a new workspace. This can be because you already have the required resources.

**Comment by ajcollins0** [comment]
@devikamehra everything is done and working as expected with reverting back to v1, and I fixed the typo's in your most recent comments. Please approve and merge when you can

**Comment by ajcollins0** [comment]
I rolled the version to be 1.0.3 now that we have a working stable build

**Comment by devikamehra** [comment]
@ajcollins0 
The latest copy of the mainTemplate.json is throwing following error in every playbook deployment.
{
    "status": "Failed",
    "error": {
        "code": "InvalidTemplate",
        "message": "Unable to process template language expressions for resource '/subscriptions/<subscription-id>/resourceGroups/<rg-name>/providers/Microsoft.Logic/workflows/CollectDefenderStatusTanium' at line '604' and column '9'. 'The template language function 'uri' expects its first argument to be a uri string. The provided value is 'https://'. Please see https://aka.ms/arm-template-expressions/#uri for usage details.'",
        "additionalInfo": [
            {
                "type": "TemplateViolation",
                "info": {
                    "lineNumber": 604,
                    "linePosition": 9,
                    "path": ""
                }
            }
        ]
    }
}

Please let me know in case I am missing something. I have tried entering the value of the host with "https://" and without.

**Comment by v-sabiraj** [comment]
@ajcollins0, can you please keep a single logo consistent and also please check the format of logo should be .svg. Please add the logo in[ logo folder](https://github.com/Azure/Azure-Sentinel/tree/master/Logos), you can refer the other logos from the same folder, logo's are still in .png format.

**Review by devikamehra — APPROVED** [review]

File: Solutions/Tanium/images/Tanium.svg
Comment (v-sabiraj) [inline Solutions/Tanium/images/Tanium.svg:1]: @ian-hands-tanium, can you please update the similar logo added in Logos/Tanium.svg, here as well, thanks.

File: Sample Data/Custom/TaniumThreatResponse_CL.csv
Comment (v-sabiraj) [inline Sample Data/Custom/TaniumThreatResponse_CL.csv:2]: Hey @ian-hands-tanium, Impact_Score, Match_Details_match_properties_file_md5, Match_Details_match_properties_md5 these columns/fields are added 2 times with same name, will you please the data by removing them from one instance if not needed, thanks.

File: Sample Data/Custom/TaniumDiscoverUnmanagedAssets_CL.csv
Comment (v-sabiraj) [inline Sample Data/Custom/TaniumDiscoverUnmanagedAssets_CL.csv:1]: @ian-hands-tanium, same is with this "OsGeneration" is added 2 times, please update the data by removing it from one instance, thanks.

**Review by aprakash13 — APPROVED** [review]
Thanks.

**Comment by v-sabiraj** [comment]
Hey @ian-hands-tanium, solution zip folder should only contain createUiDefinition and mainTemplate files, I see that some folder __MACOSX is added in that zip folder, will you please update the zip by removing that folder, thanks.

File: Sample Data/Custom/TaniumDiscoverUnmanagedAssets_CL.csv
Comment (brandonwise) [reply Sample Data/Custom/TaniumDiscoverUnmanagedAssets_CL.csv]: Thanks, @v-sabiraj for the assistance. I have made the requested changes. Please let me know if these look ok.

**Review by v-sabiraj — APPROVED** [review]

**Comment by v-sabiraj** [comment]
@ian-hands-tanium, @brandonawise , just a small change the package i.e solution version is 1.0.4 but the content version is still 1.0.3, can you please update the content version in mainTemplate and keep it consistent, thanks.
image

**Pull request merged**
